### PR TITLE
replace select with epoll

### PR DESCRIPTION
**Pull request by ggary9424**

### CODE REVIEW

**Comment by jserv** [comment]
You should mention the benefit of epoll migration such as performance gain or the ease of maintenance.

File: include/server.h [hunk 1]
```diff
@@ -1,13 +1,14 @@
 #ifndef SERVER_H
 #define SERVER_H
 
+#include <sys/epoll.h>
```
Comment (mattn) [inline include/server.h:4]: block with `#ifndef _WIN32`

File: include/server.h [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 #include "list.h"
 #include "request.h"
 #include "response.h"
 
 typedef struct Server {
     unsigned int port;
-
+    int epollfd;
```
Comment (mattn) [inline include/server.h:11]: ditto

File: src/server.c [hunk 3]
```diff
@@ -9,17 +9,21 @@
 # include <sys/select.h>
 # include <sys/types.h>
 # include <time.h>
+# include <fcntl.h>
 # include <unistd.h>
+# include <errno.h>
 typedef int sockopt_t;
 #else
 # define FD_SETSIZE 4096
+# include <fcntl.h>
```
Comment (mattn) [inline src/server.c:18]: movedown this after ws2tcpip.h

File: src/server.c [hunk 4]
```diff
@@ -64,6 +68,7 @@ Server *serverNew(unsigned int port)
 {
     Server *server = malloc(sizeof(Server));
     server->port = port;
+    server->epollfd = epoll_create1(0);
```
Comment (mattn) [inline src/server.c:71]: block with `#ifndef _WIN32`

File: src/server.c [hunk 5]
```diff
@@ -182,7 +187,34 @@ static inline int makeSocket(unsigned int port)
     return sock;
 }
 
-static inline void handle(Server *server, int fd, fd_set *activeFDs, struct sockaddr_in *addr)
+static void setFdNonblocking(int fd)
```
Comment (mattn) [inline src/server.c:190]: ditto

File: src/server.c [hunk 6]
```diff
@@ -220,10 +252,8 @@ static inline void handle(Server *server, int fd, fd_set *activeFDs, struct sock
             requestDel(req);
         }
     }
-
+    serverDelFd(server->epollfd, fd);
```
Comment (mattn) [inline src/server.c:255]: maybe you've better to pass server instead of epollfd because you need to add `#ifndef _WIN32`

File: src/server.c [hunk 7]
```diff
@@ -232,46 +262,37 @@ void serverServe(Server *server)
...
-    int newSock;
-
-    socklen_t size;
-
-    fd_set activeFDs;
-    fd_set readFDs;
-
+    int sock, newSock, tmpfd, nfds;
     struct sockaddr_in addr;
+    struct epoll_event * events = (struct epoll_event *)malloc(sizeof(struct epoll_event) * 64);
```
Comment (mattn) [inline src/server.c:267]: ditto

**Comment by jserv** [comment]
@ggary9424, would you like to rework this patch?

**Comment by jserv** [comment]
Drop this pull request since the new implementation was merged.

**Pull request closed without merging**